### PR TITLE
expose migration region

### PR DESCRIPTION
**Pull request by tmakatos**
This PR depends on some other pending PRs so I'll leave it as a draft, however it would be good to get some initial feedback.

Also, need to implement unit tests.

### CODE REVIEW

File: include/libvfio-user.h [hunk 1]
```diff
@@ -240,6 +245,25 @@ typedef ssize_t (vfu_region_access_cb_t)(vfu_ctx_t *vfu_ctx, char *buf,
  *    and writes are an error
  *  - otherwise, the callback is expected to handle the access
  *
+ * Regions VFU_PCI_DEV_MIGR_REGION_IDX and VFU_GENERIC_DEV_MIGR_REG_IDX,
+ * corresponding to the migration region, enable live migration support for
+ * the device. The migration region must contain at the beginning the migration
+ * registers (struct vfio_device_migration_info defined in <linux/vfio.h>) and
+ * the remaining part of the region can be arbitrarily used by the device
+ * implementation. Therefore, the size of the migration region must be at least
+ * sizeof(struct vfio_device_migration_info). libvfio-user offers two ways for
```
Comment (tmakatos) [inline include/libvfio-user.h:254]: Add function that returns page aligned space required for migration registers.

File: include/libvfio-user.h [hunk 2]
```diff
@@ -422,21 +446,26 @@ typedef struct {
...
-
-//TODO: Re-visit once migration support is done.
 /**
- * Enable support for device migration.
+ * vfu_setup_device_migration provides an abstraction over the migration
+ * protocol: the user specifies a set of callbacks which are called in respone
+ * to client accesses of the migration region; the migration read/write
+ * callbacks are not called. Offsets in callbacks are relative to @data_offset.
+ * 
+ * FIXME we shouldn't be using vfu_migration_callbacks_t for the same reason
```
Comment (tmakatos) [inline include/libvfio-user.h:455]: Add single function that sets migration callback based on enum.

**Review by tmakatos — COMMENTED** [review]
Added comments based on meeting.

**Comment by tmakatos** [comment]
I think configuring migration has started to become a bit cumbersome: the user needs to create the migration region (`vfu_setup_region`) making sure the size is correct (by calling the newly-added `vfu_get_migr_regs_size`) and then configure it via `vfu_setup_device_migration_callbacks`. I think it still makes sense to add another function that combines all these (allocating the migration region memory if it's to be sparse mmaped) for the benefit of simpler implementations.

**Comment by jlevon** [comment]
I don't agree that we want a new function that literally *is* adding a region, plus other stuff. It's *more* confusing to me to have it look like it's something new/different when it isn't.

**Comment by tmakatos** [comment]
> I don't agree that we want a new function that literally _is_ adding a region, plus other stuff. It's _more_ confusing to me to have it look like it's something new/different when it isn't.

Fair enough.

**Comment by tmakatos** [comment]
All the commits will be squashed since they're very relevant, I've put them separately to make reviewing easier.

File: include/libvfio-user.h [hunk 3]
```diff
@@ -240,6 +245,26 @@ typedef ssize_t (vfu_region_access_cb_t)(vfu_ctx_t *vfu_ctx, char *buf,
  *    and writes are an error
  *  - otherwise, the callback is expected to handle the access
  *
+ * Regions VFU_PCI_DEV_MIGR_REGION_IDX and VFU_GENERIC_DEV_MIGR_REG_IDX,
+ * corresponding to the migration region, enable live migration support for
+ * the device. The migration region must contain at the beginning the migration
+ * registers (struct vfio_device_migration_info defined in <linux/vfio.h>) and
+ * the remaining part of the region can be arbitrarily used by the device
+ * implementation. Therefore, the size of the migration region must be at least
+ * sizeof(struct vfio_device_migration_info). libvfio-user offers two ways for
```
Comment (jlevon) [inline include/libvfio-user.h:254]: Can we clarify this, specifically say something like:

The region provided must have at least vfu_get_migr_regs_size() bytes available at the start of the region (this size is guaranteed to be page-aligned). If mmap_areas is given, it must *not* include this part of the region.

Comment (jlevon) [inline include/libvfio-user.h:254]: New para here? it's a bit hard to follow all in one

File: include/libvfio-user.h [hunk 4]
```diff
@@ -240,6 +245,26 @@ typedef ssize_t (vfu_region_access_cb_t)(vfu_ctx_t *vfu_ctx, char *buf,
...
+ * the migration region to be used:
+ *  1. natively: the device implementation must handle accesses to the
+ *      migration registers and migration data via the region callbacks. The
+ *      semantics of these registers are explained in <linux/vfio.h>.
+ *  2. via the vfu_migration_t callbacks: the device implementation registers
+ *      a set of callbacks by calling vfu_setup_device_migration. The device
```
Comment (jlevon) [inline include/libvfio-user.h:260]: I think you can remove this "The device ... see function vfu_get_migr_regs_size)". if you have the above. Right now it sounds like if you choose 1) then you can do what you want with mmap, but that's not really true is it?

File: include/libvfio-user.h [hunk 5]
```diff
@@ -240,6 +245,26 @@ typedef ssize_t (vfu_region_access_cb_t)(vfu_ctx_t *vfu_ctx, char *buf,
...
+ *      does not have access to the migration registers, however the migration
+ *      region provided must still reserve space for them (see function
+ *      vfu_get_migr_regs_size). The region's read/write callbacks are never
+ *      called.
+ * The migration region can be partially memory mapped, except the migration
```
Comment (jlevon) [inline include/libvfio-user.h:265]: Think you can remove this too if you do the above

File: include/libvfio-user.h [hunk 6]
```diff
@@ -261,6 +286,13 @@ vfu_setup_region(vfu_ctx_t *vfu_ctx, int region_idx, size_t size,
                  struct iovec *mmap_areas, uint32_t nr_mmap_areas,
                  int fd);
 
+/*
+ * Returns the size of the migration registers, which is guaranteed to be page
```
Comment (jlevon) [inline include/libvfio-user.h:290]: This makes it sound like the registers themselves take up a page. Maybe generalize a bit?

Returns the size of the area needed to hold the migration registers at the beginning of the migration region; guaranteed to be page aligned.

and call it "vfu_get_migr_register_area_size" or somesuch? What do you think?

File: include/libvfio-user.h [hunk 7]
```diff
@@ -422,21 +461,22 @@ typedef struct {
...
-    vfu_migration_callbacks_t   callbacks;
-    struct iovec                *mmap_areas;
-    uint32_t                    nr_mmap_areas;
-} vfu_migration_t;
-
-//TODO: Re-visit once migration support is done.
 /**
- * Enable support for device migration.
+ * vfu_setup_device_migration provides an abstraction over the migration
+ * protocol: the user specifies a set of callbacks which are called in respone
```
Comment (jlevon) [inline include/libvfio-user.h:466]: nit, response

File: include/libvfio-user.h [hunk 8]
```diff
@@ -422,21 +461,22 @@ typedef struct {
...
+ * to client accesses of the migration region; the migration read/write
```
Comment (jlevon) [inline include/libvfio-user.h:467]: I think you mean the "region read/write callbacks are not called"; maybe add "after this function call".

File: include/libvfio-user.h [hunk 9]
```diff
@@ -422,21 +461,22 @@ typedef struct {
...
- * @migration: information required to migrate device
+ * @callbacks: migration callbacks
+ * @data_offset: offset in the migration region where data begins.
+ *
+ * @returns 0 on success, -1 on error, Sets errno.
  */
 int
-vfu_setup_device_migration(vfu_ctx_t *vfu_ctx, vfu_migration_t *migration);
+vfu_setup_device_migration_callbacks(vfu_ctx_t *vfu_ctx,
+                                     const vfu_migration_callbacks_t * callbacks,
```
Comment (jlevon) [inline include/libvfio-user.h:478]: nit, spurious space after *

File: include/libvfio-user.h [hunk 10]
```diff
@@ -422,21 +461,22 @@ typedef struct {
...
+ * protocol: the user specifies a set of callbacks which are called in respone
+ * to client accesses of the migration region; the migration read/write
+ * callbacks are not called. Offsets in callbacks are relative to @data_offset.
+ * 
  * @vfu_ctx: the libvfio-user context
- * @migration: information required to migrate device
+ * @callbacks: migration callbacks
+ * @data_offset: offset in the migration region where data begins.
+ *
+ * @returns 0 on success, -1 on error, Sets errno.
```
Comment (jlevon) [inline include/libvfio-user.h:474]: nit, s/Sets/sets/

File: include/libvfio-user.h [hunk 11]
```diff
@@ -550,6 +590,18 @@ vfu_dma_write(vfu_ctx_t *vfu_ctx, dma_sg_t *sg, void *data);
 
 /*
  * Supported PCI regions.
+ *
+ * Note to developers: in VFIO, each region starts at a terabyte offset
```
Comment (jlevon) [inline include/libvfio-user.h:594]: I don't think you need "Note to developers": everyone's a developer :)

File: include/libvfio-user.h [hunk 12]
```diff
@@ -550,6 +590,18 @@ vfu_dma_write(vfu_ctx_t *vfu_ctx, dma_sg_t *sg, void *data);
...
+ * (VFIO_PCI_INDEX_TO_OFFSET) and because Linux supports up to 128 TB of
+ * virtual memory, there can be up to 128 device regions. PCI regions are fixed
```
Comment (jlevon) [inline include/libvfio-user.h:596]: *userspace* virtual memory ?

File: include/libvfio-user.h [hunk 13]
```diff
@@ -550,6 +590,18 @@ vfu_dma_write(vfu_ctx_t *vfu_ctx, dma_sg_t *sg, void *data);
...
+ * and in retrospect this choice has proven to be problematic because devices
+ * might contain potentially unused regions. New regions can now be positioned
+ * anywhere by using the VFIO_REGION_INFO_CAP_TYPE capability.
+ * In vfio-user we don't have this problem because the region index is just an
+ * identifier: the VMM memory maps a file descriptor that is passed to it and
+ * the offset that it memory maps is relative to the beginning of the region,
```
Comment (jlevon) [inline include/libvfio-user.h:602]: I think the crucial point to make here is that the mapping offset is derived from mmap_areas offset value, rather than a static mapping from region index to offset.

File: lib/libvfio-user.c [hunk 14]
```diff
@@ -1234,6 +1239,38 @@ copyin_mmap_areas(vfu_reg_info_t *reg_info,
...
+     * utility functions?
+     */
+    return (off1 < (off2 + size2) && off2 < (off1 + size1));
+}
+
+static bool
+maps_over_migr_regs(struct iovec *iov)
+{
+    return ranges_intersect(0, sizeof(struct vfio_device_migration_info),
+                            (size_t)iov->iov_base, (size_t)iov->iov_len);
```
Comment (jlevon) [inline lib/libvfio-user.c:1258]: nit, no need to cast iov_len

File: lib/libvfio-user.c [hunk 15]
```diff
@@ -1266,6 +1303,12 @@ vfu_setup_region(vfu_ctx_t *vfu_ctx, int region_idx, size_t size,
         return ERROR(EINVAL);
     }
 
+    if (region_idx == VFU_PCI_DEV_MIGR_REGION_IDX &&
+        size < sizeof(struct vfio_device_migration_info)) {
+        vfu_log(vfu_ctx, LOG_ERR, "ivalid migration region size %d", size);
```
Comment (jlevon) [inline lib/libvfio-user.c:1308]: nit, "invalid"

File: lib/libvfio-user.c [hunk 16]
```diff
@@ -1266,6 +1303,12 @@ vfu_setup_region(vfu_ctx_t *vfu_ctx, int region_idx, size_t size,
         return ERROR(EINVAL);
     }
 
+    if (region_idx == VFU_PCI_DEV_MIGR_REGION_IDX &&
+        size < sizeof(struct vfio_device_migration_info)) {
```
Comment (jlevon) [inline lib/libvfio-user.c:1307]: This should be requiring the page aligned size (i.e. what your utility function returns), not the struct size. Otherwise it's just more confusing I think.

File: lib/libvfio-user.c [hunk 17]
```diff
@@ -1344,41 +1406,39 @@ vfu_setup_device_nr_irqs(vfu_ctx_t *vfu_ctx, enum vfu_dev_irq_type type,
...
+vfu_get_migr_regs_size(void)
+{
+    return ROUND_UP(sizeof(struct vfio_device_migration_info),
+                    sysconf(_SC_PAGE_SIZE));
+}
+
 int
-vfu_setup_device_migration(vfu_ctx_t *vfu_ctx, vfu_migration_t *migration)
+vfu_setup_device_migration_callbacks(vfu_ctx_t *vfu_ctx,
+                                     const vfu_migration_callbacks_t * callbacks,
```
Comment (jlevon) [inline lib/libvfio-user.c:1418]: nit, spurious space after *

File: lib/migration.c [hunk 18]
```diff
@@ -51,9 +51,14 @@ enum migr_iter_state {
 };
 
 struct migration {
+    /*
+     * TODO if the user provides an FD then should mmap it an use the migration
```
Comment (jlevon) [inline lib/migration.c:55]: nit, "and use"

File: lib/migration.h [hunk 19]
```diff
@@ -44,7 +44,8 @@
 #include "libvfio-user.h"
 
 struct migration *
-init_migration(const vfu_migration_t * const vfu_migr, int *err);
+init_migration(const vfu_migration_callbacks_t * callbacks,
```
Comment (jlevon) [inline lib/migration.h:47]: nit, spurious space after *

File: samples/client.c [hunk 20]
```diff
@@ -217,12 +217,13 @@ get_region_vfio_caps(struct vfio_info_cap_header *header,
     while (true) {
         switch (header->id) {
             case VFIO_REGION_INFO_CAP_SPARSE_MMAP:
-                sparse = (struct vfio_region_info_cap_sparse_mmap*)header;
+                *sparse = (struct vfio_region_info_cap_sparse_mmap*)header;
```
Comment (jlevon) [inline samples/client.c:220]: missing space before *

File: samples/server.c [hunk 21]
```diff
@@ -501,21 +523,41 @@ int main(int argc, char *argv[])
...
+        err(EXIT_FAILURE, "failed to truncate migration file");
+    }
+    server_data.bar1 = mmap(NULL, server_data.bar1_size, PROT_READ | PROT_WRITE,
+                            MAP_SHARED, fileno(bar1_fp), 0);
+    if (server_data.bar1 == MAP_FAILED) {
+        err(EXIT_FAILURE, "failed to mmap migration file");
+    }
+    struct iovec migr_mmap_areas[] = {
+        [0] = {
+            .iov_base  = (void*)migr_regs_size,
```
Comment (jlevon) [inline samples/server.c:549]: nit, missing space before *

File: test/unit-tests.c [hunk 22]
```diff
@@ -1202,11 +1203,157 @@ test_migration_state_transitions(void **state __attribute__ ((unused)))
...
+static struct test_setup_migr_reg_dat {
+    vfu_ctx_t *v;
+    size_t rs; /* migration registers size */
+    size_t ds; /* migration data size */
+    size_t s; /* migration region size*/
+    const vfu_migration_callbacks_t c;
+} migr_reg_data = {
+    .c = {
+        .version = VFU_MIGR_CALLBACKS_VERS,
+        .transition = (void*)0x1,
```
Comment (jlevon) [inline test/unit-tests.c:1221]: nit, here and below, void* -> void *

File: test/unit-tests.c [hunk 23]
```diff
@@ -1202,11 +1203,157 @@ test_migration_state_transitions(void **state __attribute__ ((unused)))
...
+        .get_pending_bytes = (void*)0x2,
+        .prepare_data = (void*)0x3,
+        .read_data = (void*)0x4,
+        .write_data = (void*)0x5,
+        .data_written = (void*)0x6
+    }
+};
+
+static int
+setup_test_setup_migration_region(void **state) {
```
Comment (jlevon) [inline test/unit-tests.c:1231]: nit, { on newline

File: test/unit-tests.c [hunk 24]
```diff
@@ -1202,11 +1203,157 @@ test_migration_state_transitions(void **state __attribute__ ((unused)))
...
+    p->ds = sysconf(_SC_PAGE_SIZE);
+    p->s = p->rs + p->ds;
+    *state = p;
+    return setup(state);
+}
+
+static vfu_ctx_t *
+get_vfu_ctx(void **state)
+{
+    return (*((struct test_setup_migr_reg_dat**)(state)))->v;
```
Comment (jlevon) [inline test/unit-tests.c:1248]: nit, s/**/ **/

File: test/unit-tests.c [hunk 25]
```diff
@@ -1202,11 +1203,157 @@ test_migration_state_transitions(void **state __attribute__ ((unused)))
...
+}
+
+static int
+teardown_test_setup_migration_retion(void **state)
```
Comment (jlevon) [inline test/unit-tests.c:1252]: nit, "region" ?

File: test/unit-tests.c [hunk 26]
```diff
@@ -1202,11 +1203,157 @@ test_migration_state_transitions(void **state __attribute__ ((unused)))
...
+test_setup_migration_region_size_ok(void **state)
+{
+    vfu_ctx_t *v = get_vfu_ctx(state);
+    int r = vfu_setup_region(v, VFU_PCI_DEV_MIGR_REGION_IDX,
+        sizeof(struct vfio_device_migration_info), NULL,
+        VFU_REGION_FLAG_READ | VFU_REGION_FLAG_WRITE, NULL, 0, -1);
+    assert_int_equal(0, r);
+}
+
+static void
```
Comment (jlevon) [inline test/unit-tests.c:1280]: I'm wondering if we should handle this ourselves. Not sure

File: test/unit-tests.c [hunk 27]
```diff
@@ -1202,11 +1203,157 @@ test_migration_state_transitions(void **state __attribute__ ((unused)))
...
+    assert_int_equal(EINVAL, errno);
+}
+
+static void
+test_setup_migration_region_sparsely_mappable_valid(void **state)
+{
+    struct test_setup_migr_reg_dat *p = *state;
+    struct iovec mmap_areas[] = {
+        [0] = {
+            .iov_base = (void*)p->rs,
```
Comment (jlevon) [inline test/unit-tests.c:1313]: nit, void *

**Review by jlevon — CHANGES_REQUESTED** [review]

File: include/libvfio-user.h (hunk 3 above)
Comment (tmakatos) [reply include/libvfio-user.h:254]: Sure, I was hoping for a better wording :).

Comment (tmakatos) [reply include/libvfio-user.h:254]: ok

File: include/libvfio-user.h (hunk 4 above)
Comment (tmakatos) [reply include/libvfio-user.h:260]: ok

File: include/libvfio-user.h (hunk 5 above)
Comment (tmakatos) [reply include/libvfio-user.h:265]: ok

File: include/libvfio-user.h (hunk 6 above)
Comment (tmakatos) [reply include/libvfio-user.h:290]: ok

File: test/unit-tests.c (hunk 23 above)
Comment (tmakatos) [reply test/unit-tests.c:1231]: ok

File: test/unit-tests.c (hunk 24 above)
Comment (tmakatos) [reply test/unit-tests.c:1248]: ok

File: test/unit-tests.c (hunk 25 above)
Comment (tmakatos) [reply test/unit-tests.c:1252]: ok

Comment (tmakatos) [reply test/unit-tests.c:1252]: 'region' isn't that accurate, since the pointer points to `struct test_setup_migr_reg_dat`, which is 'migration region test data'. Honestly I think it'f fine as 'state' since the cast is right below.

File: test/unit-tests.c (hunk 26 above)
Comment (tmakatos) [reply test/unit-tests.c:1280]: Do you mean making the migration region mappable without telling the user?

File: test/unit-tests.c (hunk 27 above)
Comment (tmakatos) [reply test/unit-tests.c:1313]: ok

File: test/unit-tests.c (hunk 25 above)
Comment (jlevon) [reply test/unit-tests.c:1252]: It's the "retion" part I was talking about

File: test/unit-tests.c (hunk 26 above)
Comment (jlevon) [reply test/unit-tests.c:1280]: So for non-migration regions, if an fd is provided, but no mmap areas, we map the whole region.
I'm saying that for the migration region, if an fd is provided , but no mmap areas, could we map the whole region, excluding the first page?

File: test/unit-tests.c (hunk 25 above)
Comment (tmakatos) [reply test/unit-tests.c:1253]: ah yes

File: test/unit-tests.c (hunk 26 above)
Comment (tmakatos) [reply test/unit-tests.c:1281]: We can't assume that the space after the migration registers is simply padding, the user might setup the migration region without calling `vfu_get_migr_register_area_size` and putting data right after the migration registers.

**Comment by tmakatos** [comment]
@jlevon fixed typo and replied to your comment regarding implicitly making the migration region mappable, please take another look.

File: test/unit-tests.c (hunk 26 above)
Comment (jlevon) [reply test/unit-tests.c:1281]: I'm not sure why we need to support that (but maybe we do?). Isn't it simpler to have a simple "page size" requirement?

Comment (tmakatos) [reply test/unit-tests.c:1281]: We don't necessarily need to support that, but seems that we're making it more complicated (a) by introducing a requirement, and (b) by changing the semantics of the how the function works if mmap areas is NULL but only if this is the migration region. I think it will be harder to get it right this way.

Comment (jlevon) [reply test/unit-tests.c:1281]: ok, but I think we should always insist on a page-sized buffer, regardless.

Comment (tmakatos) [reply test/unit-tests.c:1281]: > ok, but I think we should always insist on a page-sized buffer, regardless.

We sure can, but if we're introducing a restriction we should explain why?

I think if we want to make the API easier we can explain that the simplest way to make the migration region mappable is for the client to use `vfu_get_migr_register_area_size` in order to allocate space for the migration registers and then

Comment (jlevon) [reply test/unit-tests.c:1281]: Right, I think we're agreeing: we should say that regardless of whether you're using the callbacks, and regardless of if you're providing an fd, you need to reserve a non-mappable area of size vfu_get_migr_register_area_size().

Comment (tmakatos) [reply test/unit-tests.c:1281]: I re-read the PR and it's currently implemented the way you suggest:
```
The migration region must contain at the beginning the migration
registers (struct vfio_device_migration_info defined in <linux/vfio.h>) and
the remaining part of the region can be arbitrarily used by the device
implementation. The region provided must have at least
vfu_get_migr_register_area_size() bytes available at the start of the region
(this size is guaranteed to be page-aligned).
```
So going back to your original comment in this discussion, we're deciding _not_ to special case the `mmap_areas == NULL` case for the migration region, right?

Comment (jlevon) [reply test/unit-tests.c:1281]: We won't special case.

But it's not implemented as you suggest, yet:

```
   return ranges_intersect(0, sizeof(struct vfio_device_migration_info),
                            (size_t)iov->iov_base, (size_t)iov->iov_len);
```

```
    if (region_idx == VFU_PCI_DEV_MIGR_REGION_IDX &&
        size < sizeof(struct vfio_device_migration_info)) {
```

At least both of these are using the struct size not the vfu_get_migr_register_area_size()

Comment (tmakatos) [reply test/unit-tests.c:1281]: Right, this needs fixing.

**Comment by tmakatos** [comment]
@jlevon fixed, please take another look.

**Comment by jlevon** [comment]
I'm not seeing changes for my review comments

**Comment by tmakatos** [comment]
> I'm not seeing changes for my review comments

I somehow missed some of them, let me take a look.

File: include/libvfio-user.h (hunk 13 above)
Comment (tmakatos) [reply include/libvfio-user.h:600]: Thanks, I know the way I wrote it was a bit convoluted and was hoping in a suggestion :).

**Comment by tmakatos** [comment]
@jlevon addressed your comments.

File: lib/irq.c [hunk 28]
```diff
@@ -427,10 +428,10 @@ vfu_irq_message(vfu_ctx_t *vfu_ctx, uint32_t subindex)
     }
 
     irq_info.subindex = subindex;
-    ret = tran_sock_msg(vfu_ctx->conn_fd, msg_id,
-                        VFIO_USER_VM_INTERRUPT,
-                        &irq_info, sizeof irq_info,
-                        NULL, NULL, 0);
+    ret = vfu_ctx->trans->send_msg(vfu_ctx, msg_id,
```
Comment (jlevon) [inline lib/irq.c:431]: I compared against master, and I'm so confused as to why github thinks there's a change here.

**Review by jlevon — APPROVED** [review]
All looks good, thanks. I'm a bit worried about why my changes are showing up in the diffs though, wtf?

File: lib/irq.c (hunk 28 above)
Comment (tmakatos) [reply lib/irq.c:431]: Not sure either, I also diff'ed against master and don't see your changes. I'll merge and if it breaks things I'll revert the merge.

**Comment by tmakatos** [comment]
@jlevon your changes don't appear in the commit on master, we're good.
